helix-core/src/main/java/org/apache/helix/messaging/handling/HelixTaskExecutor.java

```diff
@@ -821,32 +843,29 @@ public void onMessage(String instanceName, List<Message> messages,
         LOG.error(
             "Exception happens when creating Message Handler for message {}. Current remaining retry count is {}.",
             message.getMsgId(), remainingRetryCount);
-        // Set the message retry count to avoid infinite retrying.
+        // Reduce the message retry count to avoid infinite retrying.
         message.setRetryCount(remainingRetryCount - 1);
         message.setExecuteSessionId(sessionId);
-        // continue processing in the next section where handler object is double-checked.
-      }
-
-      if (msgHandler == null) {
         // Note that we are re-using the retry count of Message that was original designed to control
         // timeout retries. So it is not checked before the first try in order to ensure consistent
         // behavior. It is possible that we introduce a new behavior for this method. But it requires
         // us to split the configuration item so as to avoid confusion.
-        if (message.getRetryCount() < 0) {
+        if (message.getRetryCount() <= 0) {
           // If no more retry count remains, then mark the message to be UNPROCESSABLE.
-          String errorMsg = String
-              .format("No available message Handler found!"
-                      + " Stop processing message %s since it has a negative remaining retry count %d!",
-                  message.getMsgId(), message.getRetryCount());
+          String errorMsg = String.format("No available message Handler found!"
+                  + " Stop processing message %s since it has negative remaining retry count %d!",
```

Comment:
"negative" doesn't match if condition.

Reply:
Good catch, I forget to change the comment after the code change.